### PR TITLE
Define protobufs for topk and cms.

### PR DESCRIPTION
**Pull request by jeschkies**
**What this PR does / why we need it**:
This change introduces Protobuf models for the sketch data structures which will be used to shard topk queries.

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)

### CODE REVIEW

File: pkg/util/sketch/series.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+// Type implements `promql.Value` and `parser.Value`
+func (Series) Type() parser.ValueType { return ValueTypeSketchSeries }
+
+// String implements `promql.Value` and `parser.Value`
+func (Series) String() string {
+	return ""
+}
+
+func (s Series) ToProto() (*logproto.SketchSeries, error) {
```
Comment (jeschkies) [inline pkg/util/sketch/series.go:28]: I wasn't sure where to put this but this seemed to be the cleanest.

File: pkg/util/sketch/series.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+			cms.counters = append(cms.counters, point.Topk.Cms.Counters[s:e])
+		}
+
+		hll := hyperloglog.New()
+		err := hll.UnmarshalBinary(point.Topk.Hyperloglog)
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO(karsten): should we set expected cardinality as well?
```
Comment (jeschkies) [inline pkg/util/sketch/series.go:75]: @cstyan how should we reconstruct the heap? Should we leave it empty and only set the labels?

File: pkg/logproto/sketch.proto [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  repeated uint32 counters = 3;
+}
+
+message TopK {
+  CountMinSketch cms = 1;
+  repeated string labels = 2;
+  bytes hyperloglog = 3;
+}
+
+message SketchSeries {
```
Comment (owen-d) [inline pkg/logproto/sketch.proto:21]: Maybe name this something like `TopKMatrix`? There will be many different sketch types. Roughly using the same `Vector, Matrix` semantics that prometheus uses where `Vector` is a list of points for one timestamp and `Matrix` is a list of vectors.

File: pkg/logproto/sketch.proto [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  // TODO(karsten): Open for a different name.
+
+  message Point {
```
Comment (owen-d) [inline pkg/logproto/sketch.proto:24]: Maybe name this `TopKVector`? Or just `Vector` if it's private to this sketch?

File: pkg/logproto/sketch.proto (hunk 3 above)
Comment (jeschkies) [reply pkg/logproto/sketch.proto:21]: I like that. I was unsure about Matrix and Vector and didn't want to cause any confusion.

File: pkg/util/sketch/series.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+// Type implements `promql.Value` and `parser.Value`
+func (TopKMatrix) Type() parser.ValueType { return ValueTypeTopKMatrix }
+
+// String implements `promql.Value` and `parser.Value`
+func (TopKMatrix) String() string {
+	return ""
+}
+
+func (s TopKMatrix) ToProto() (*logproto.TopKMatrix, error) {
```
Comment (owen-d) [inline pkg/util/sketch/series.go:28]: Is there a way we can align the proto structure and the logqlmodel structure? That would let us avoid doing all this translation which would be faster & have less code to maintain.

**Review by owen-d — COMMENTED** [review]
A question but LGTM

File: pkg/util/sketch/series.go (hunk 5 above)
Comment (jeschkies) [reply pkg/util/sketch/series.go:28]: The original version did this. However, I didn't want the encoding to impact the performance of algorithm. Eg we might decide on a sparse matrix encoding that would not be fast for the access pattern of the algorithm. 

I'm open to change it back though.

File: pkg/util/sketch/series.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+		labels := make([]string, len(*point.topk.heap))
+		for _, node := range *point.topk.heap {
+			labels = append(labels, node.event)
+		}
+
+		topk := &logproto.TopK{
+			Cms:         cms,
+			Hyperloglog: hllBytes,
+			Labels:      labels,
+		}
```
[the remark this answers is not in the document]
Comment (jeschkies) [reply pkg/util/sketch/series.go]: Done. Could you take another look?

**Review by dannykopping — APPROVED** [review]
Don't have much to add to what's been discussed, approving!

**Pull request closed without merging**